lib/rubygems/text.rb

```diff
@@ -86,7 +86,7 @@ def levenshtein_distance(str1, str2)
 end
 
 # Use the latest and greatest DidYouMean::Levenshtein if it is available
-begin
+if RUBY_VERSION >= '2.7' && defined?(DidYouMean)
```

Comment:
In my testing in `irb` as well as `ruby -e` on ruby 2.3 to 2.7 I get

`defined?(DidYouMean)` | `defined?(DidYouMean::Levenshtein)`
--- | ---
"constant" | "constant"

However, in the context of the rubygems runner as tested with `rake install` or `ruby -Ilib bin/gem ....` I get 

`defined?(DidYouMean)` | `defined?(DidYouMean::Levenshtein)`
--- | ---
"constant" | nil

I didn't get to the root of it, but I can definitely see that rubygems tries to isolate it's environment for good reason. On ruby 2.6 downwards in the context of the rubygems runner: the bundled did_you_mean gem is not in the load path so we cannot require `did_you_mean/levenshtein`. On ruby 2.7 upwards `did_you_mean/levenshtein` is always available for requiring.

**I like this better anyway:** it simplifies the logic here:

_note that the vendored version is based on did_you_mean 1.4.0_


. | default | --disable-did_you_mean
--- | --- | ---
**ruby >=2.7** | did_youmean >= 1.4.0 | use vendored
**ruby <2.7** | use vendored | use vendored

**:heavy_check_mark: Simple**

Compare this with if we were to somehow successfully use the bundled version on ruby <2.7:

. | default | --disable-did_you_mean
--- | --- | ---
**ruby >=2.7** | did_youmean >= 1.4.0 | use vendored
**ruby 2.6** | did_you_mean 1.3.0 | use vendored
**ruby 2.5** | did_you_mean 1.2.0 | use vendored
**ruby 2.4** | did_you_mean 1.1.0 | use vendored
**ruby 2.3** | did_you_mean 1.0.0 | use vendored

**:unamused: Complicated, inconsistent performance optimizations**

Reply:
Here's some code to test that `defined?(DidYouMean::Levenshtein)` matrix. Requires docker

https://gist.github.com/austinpray/31972f4364d3269d511e1b478147492c

Reply:
I see, that makes sense.

I wonder though, is it so valuable to use `did_you_mean`'s version and not just always use the vendored version?
I think that would simplify things a lot more and make it fully reliable (e.g., `did_you_mean` renames the file in a future release and RubyGems will break (!), bad and unexpected).
I think it's best RubyGems avoids any unneeded dependency.

Reply:
@eregon Yeah, that sounds fine to me. I would rather not cause headaches down the line. I went ahead and dropped the DidYouMean dependency. I can take any further ideas of using DidYouMean to improve this feature into a different PR.

The benchmarks still hold that the updated function is a 2x speedup raw iterations-wise. In unscientific testing it shaves about a second off of the suggestion time on my machine:
![Screen Capture_select-area_20200801102036](https://user-images.githubusercontent.com/2192970/89104903-4ca25c80-d3e2-11ea-90d7-50ecc0f640ae.png)